dialer.go

```diff
@@ -50,97 +62,127 @@ func NewDialer(opts *DialerOpts) Dialer {
-		maxStreamPerConn: opts.MaxStreamsPerConn,
-		idleInterval:     opts.IdleInterval,
-		pingInterval:     opts.PingInterval,
-		pool:             opts.Pool,
-		cipherCode:       opts.Cipher,
-		serverPublicKey:  opts.ServerPublicKey,
+		windowSize:            opts.WindowSize,
+		maxPadding:            opts.MaxPadding,
+		maxStreamsPerConn:     opts.MaxStreamsPerConn,
+		maxLiveConns:          opts.MaxLiveConns,
+		idleInterval:          opts.IdleInterval,
+		pingInterval:          opts.PingInterval,
+		redialSessionInterval: opts.RedialSessionInterval,
+		pool:                  opts.Pool,
+		cipherCode:            opts.Cipher,
+		serverPublicKey:       opts.ServerPublicKey,
+		liveSessions:          liveSessions,
+		numLivePending:        1, // the nullSession
 	}
 }
 
 type dialer struct {
-	windowSize       int
-	maxPadding       int
-	maxStreamPerConn uint16
-	idleInterval     time.Duration
-	pingInterval     time.Duration
-	pool             BufferPool
-	cipherCode       Cipher
-	serverPublicKey  *rsa.PublicKey
-	current          *session
-	lastDialed       time.Time
-	id               uint16
-	mx               sync.Mutex
+	windowSize            int
+	maxPadding            int
+	maxLiveConns          int
+	maxStreamsPerConn     uint16
+	idleInterval          time.Duration
+	pingInterval          time.Duration
+	redialSessionInterval time.Duration
+	pool                  BufferPool
+	cipherCode            Cipher
+	serverPublicKey       *rsa.PublicKey
+	muNumLivePending      sync.Mutex
+	numLivePending        int
+	liveSessions          chan sessionInf
 }
 
 func (d *dialer) Dial(dial DialFN) (net.Conn, error) {
-	return d.DialStream(dial)
+	return d.DialContext(context.Background(), dial)
 }
 
-func (d *dialer) DialStream(dial DialFN) (Stream, error) {
-	d.mx.Lock()
-	current := d.current
-	idsExhausted := false
-	if d.id > d.maxStreamPerConn {
-		log.Debug("Exhausted maximum allowed IDs on one physical connection, will open new connection")
-		idsExhausted = true
-		d.id = 0
-	}
-	idled := false
-	if d.idleInterval > 0 {
-		now := time.Now()
-		idled = now.Sub(d.lastDialed) > d.idleInterval
-		if idled {
-			log.Debugf("No new connections in %v, will start new session", d.idleInterval)
-		}
-		d.lastDialed = now
+func (d *dialer) DialContext(ctx context.Context, dial DialFN) (net.Conn, error) {
+	s, err := d.getOrCreateSession(ctx, dial)
+	if err != nil {
+		return nil, err
 	}
+	c := s.CreateStream()
+	d.liveSessions <- s
+	return c, nil
+}
 
-	// TODO: support pooling of connections (i.e. keep multiple physical connections in flight)
-	if current == nil || idsExhausted || idled {
-		if current != nil {
-			// Need to run this on a goroutine because we call back to sessionClosed()
-			// via the session's beforeClose callback, which needs the mutex that's
-			// already locked here.
-			go current.markDefunct()
-		}
+func (d *dialer) getNumLivePending() int {
+	d.muNumLivePending.Lock()
+	numLivePending := d.numLivePending
```

Comment:
I tried to use `sync/atomic` package to mange `numLivePending`, but the tests turned out to fail randomly because of the order of statements, so I changed to this safer method.